shared/subspace-data-retrieval/src/object_fetcher.rs

```diff
@@ -0,0 +1,699 @@
+    }
+
+    /// Fetch and assemble an object that can cross segment boundaries, which requires assembling
+    /// and iterating over full segments.
+    async fn fetch_object_regular(
+        &self,
+        piece_index: PieceIndex,
+        piece_offset: u32,
+    ) -> Result<Vec<u8>, Error> {
+        let segment_index = piece_index.segment_index();
+        let piece_position_in_segment = piece_index.position();
+        // Used to access the data after it is converted to raw bytes
+        let offset_in_segment =
+            piece_position_in_segment as usize * RawRecord::SIZE + piece_offset as usize;
+
+        let mut data = {
+            let Segment::V0 { items } = self
+                .read_segment(segment_index, piece_index, piece_offset)
+                .await?;
+            // Go through the segment until we reach the offset.
+            // Unconditional progress is enum variant + compact encoding of number of elements
+            let mut progress = 1 + Compact::compact_len(&(items.len() as u64));
+            let segment_item = items
+                .into_iter()
+                .find(|item| {
+                    // Add number of bytes in encoded version of segment item
+                    progress += item.encoded_size();
+
+                    // Our data is within another segment item, which will have wrapping data
+                    // structure, hence strictly `>` here
+                    progress > offset_in_segment
+                })
+                .ok_or_else(|| {
+                    debug!(
+                        progress,
+                        offset_in_segment,
+                        ?segment_index,
+                        %piece_index,
+                        piece_offset,
+                        "Failed to find item at offset in segment"
+                    );
+
+                    Error::NoSegmentItem {
+                        progress,
+                        offset_in_segment,
+                        segment_index,
+                        piece_index,
+                        piece_offset,
+                    }
+                })?;
+
+            // Look at the item after the offset, collecting block bytes
+            match segment_item {
+                SegmentItem::Block { bytes, .. }
+                | SegmentItem::BlockStart { bytes, .. }
+                | SegmentItem::BlockContinuation { bytes, .. } => {
+                    // Rewind back progress to the beginning of the number of bytes
+                    progress -= bytes.len();
+                    // Get a chunk of the bytes starting at the position we care about
+                    Vec::from(&bytes[offset_in_segment - progress..])
+                }
+                segment_item @ SegmentItem::Padding
+                | segment_item @ SegmentItem::ParentSegmentHeader(_) => {
+                    // TODO: create a Display impl for SegmentItem that is shorter than the entire
+                    // data contained in it
```

Comment:
If we return an error, I don't see the point of printing anything at all. Caller will likely also print it or show in some other way to the user, which means we'll be printing the same information twice, which isn't very helpful for debugging.

Reply:
The TODO is about making displaying that data to the user easier, wherever it is printed.

The debug log below isn't shown by default, because the default log level is `info`. The `debug` shows all available data, but the caller might just print the piece index and offset (the rest can be derived).

No changes here.